In [2]:
import pandas as pd
import importlib
import gc
import sys
sys.path.append('./scripts')  
import target_pcp
importlib.reload(target_pcp)

<module 'target_pcp' from 'c:\\Users\\Usuario\\Documents\\Universidad\\austral\\2025\\Lab3\\Lab3-MCD\\notebooks\\entregable\\./scripts\\target_pcp.py'>

In [3]:
df = pd.read_csv('./datasets/periodo_x_cliente_x_producto.csv', sep=',', encoding='utf-8')
df.shape

MemoryError: Unable to allocate 782. MiB for an array with shape (6, 17082581) and data type float64

In [4]:
##### Dedicimos tomar como target TN+2
df = target_pcp.target_tn_mas_dos(df)
df

product_id  customer_id  periodo  nacimiento_producto  \
0              20001        10234   201701               201701   
1              20001        10032   201701               201701   
2              20001        10217   201701               201701   
3              20001        10125   201701               201701   
4              20001        10012   201701               201701   
...              ...          ...      ...                  ...   
15487660       21281        10212   201708               201702   
15487661       21281        10376   201708               201702   
15487662       21281        10576   201708               201702   
15487663       21281        10207   201708               201702   
15487664       21281        10630   201708               201702   

          muerte_producto  nacimiento_cliente  muerte_cliente  mes_n  \
0                  201912              201701          201912      1   
1                  201912              201701          201912      2   
2                  201912              201701          201912      3   
3                  201912              201701          201911      4   
4                  201912              201701          201912      5   
...                   ...                 ...             ...    ...   
15487660           201708              201708          201912   3500   
15487661           201708              201708          201912   3501   
15487662           201708              201708          201911   3502   
15487663           201708              201708          201912   3503   
15487664           201708              201708          201709   3504   

          total_meses  producto_nuevo  ...         cat2     cat3  brand  \
0               17776               0  ...  ROPA LAVADO  Liquido  ARIEL   
1               17776               0  ...  ROPA LAVADO  Liquido  ARIEL   
2               17776               0  ...  ROPA LAVADO  Liquido  ARIEL   
3               17776               0  ...  ROPA LAVADO  Liquido  ARIEL   
4               17776               0  ...  ROPA LAVADO  Liquido  ARIEL   
...               ...             ...  ...          ...      ...    ...   
15487660         3504               0  ...          NaN      NaN    NaN   
15487661         3504               0  ...          NaN      NaN    NaN   
15487662         3504               0  ...          NaN      NaN    NaN   
15487663         3504               0  ...          NaN      NaN    NaN   
15487664         3504               0  ...          NaN      NaN    NaN   

         sku_size stock_final        tn  plan_precios_cuidados  \
0          3000.0         NaN   0.33579                    0.0   
1          3000.0         NaN  12.31230                    0.0   
2          3000.0         NaN   0.00000                    NaN   
3          3000.0         NaN   0.08954                    0.0   
4          3000.0         NaN   6.97324                    0.0   
...           ...         ...       ...                    ...   
15487660      NaN         NaN   0.00000                    NaN   
15487661      NaN         NaN   0.00000                    NaN   
15487662      NaN         NaN   0.00000                    NaN   
15487663      NaN         NaN   0.00000                    NaN   
15487664      NaN         NaN   0.00000                    NaN   

          cust_request_qty  cust_request_tn     target  
0                      1.0          0.33579    0.00000  
1                      2.0         12.31230   15.39038  
2                      NaN              NaN    0.00000  
3                      1.0          0.08954    0.10074  
4                      5.0          6.97324  195.14996  
...                    ...              ...        ...  
15487660               NaN              NaN        NaN  
15487661               NaN              NaN        NaN  
15487662               NaN              NaN        NaN  
15487663               NaN              NaN        NaN  
15487664           

In [5]:
df[(df['product_id'] == 20002) & (df['customer_id'] == 10015)][['periodo', 'product_id', 'tn', 'target']]

periodo  product_id         tn     target
17782   201701       20002   13.39311   29.78130
18215   201702       20002   24.94396   28.02946
18697   201703       20002   29.78130   39.78374
19197   201704       20002   28.02946   47.24320
19699   201705       20002   39.78374   12.43242
20202   201706       20002   47.24320   14.91890
20707   201707       20002   12.43242   52.21616
21211   201708       20002   14.91890   52.21616
21719   201709       20002   52.21616   12.88658
22230   201710       20002   52.21616   25.17764
22734   201711       20002   12.88658   58.92607
23238   201712       20002   25.17764   18.07959
23739   201801       20002   58.92607   37.16360
24237   201802       20002   18.07959   17.68898
24731   201803       20002   37.16360   27.62159
25226   201804       20002   17.68898   42.96692
25719   201805       20002   27.62159   24.79805
26215   201806       20002   42.96692   76.09052
26715   201807       20002   24.79805   34.34006
27216   201808       20002   76.09052    0.00000
27715   201809       20002   34.34006   49.45103
28213   201810       20002    0.00000   15.92567
28711   201811       20002   49.45103   52.70982
29208   201812       20002   15.92567   51.06926
29704   201901       20002   52.70982   43.11201
30200   201902       20002   51.06926   49.10506
30696   201903       20002   43.11201   55.24319
31196   201904       20002   49.10506   37.90018
31695   201905       20002   55.24319   55.36595
32194   201906       20002   37.90018  110.48638
32687   201907       20002   55.36595    0.61381
33180   201908       20002  110.48638   46.51589
33668   201909       20002    0.61381   10.22278
34153   201910       20002   46.51589   68.61315
34645   201911       20002   10.22278        NaN
35129   201912       20002   68.61315        NaN

Obviamente, el target de los periodos 201911 y 201912 serán nulos.

In [5]:
# df['target'] = df['target'].fillna(0)

Y decidimos llenarlos con nulos, porque al entrenar el target no puede ser cero.

In [6]:
df.to_csv("./datasets/periodo_x_cliente_xproducto_con_target.csv", index=False, sep=',', encoding='utf-8')